Rewrite coreos-installer in Rust

bgilbert opened the pull request:
Features:

- Checks that target disk is not in use
- Obtains image from local file, remote URL, or stream metadata
- Streaming fetch and decompression
- Compression format sniffing
- Progress reporting
- GPG signature verification, mandatory unless `--insecure` specified
- Does not write partition table until signature is verified
- Can copy Ignition config to disk
- Can override Ignition platform ID
- Can write first-boot kernel arguments
- Doesn't depend on uniqueness of `/boot` UUID
- Wipes destination partition table on error
- Only executable dependencies: `gpg`, `lsblk`, `udevadm`

Command-line parameters are not compatible with the original, but the intent is that the `coreos.inst.*` kargs, to be implemented in a separate wrapper script, will maintain compatibility.

lucab commented:
It would be useful to turn on travis on this repo, before merging this PR. I don't seem to have the powers to do that myself though.

lucab commented on src/main.rs:
```diff
@@ -0,0 +1,464 @@
+    Ok(())
+}
+
+// Parse command-line arguments.
```
Here and pretty much everywhere else in the codebase: docstrings require three slashes `///`.

bgilbert commented on src/main.rs:
```diff
@@ -0,0 +1,464 @@
+    // format sniffing.  Don't trust the content-type since the server
+    // may not be configured correctly, or the file might be local.
+    // Then wrap in a reader for decompression.
+    let mut buf_reader = BufReader::with_capacity(16384, &mut progress_reader);
```
No particular reason.  I didn't want to implicitly trust that the default buffer size is large enough for the largest magic number.  I've switched to the default buffer with an assert.

bgilbert commented on src/main.rs:
```diff
@@ -0,0 +1,464 @@
+        }
+    };
+
+    // Cache the first MB of input and write zeroes instead.  This ensures
```
The first MiB should cover all the disk headers because it's usually where the first partition starts on modern systems (x86 at least).  It's probably a larger buffer than necessary.

Credit to @imcleod for the idea!

bgilbert commented on src/main.rs:
```diff
@@ -0,0 +1,464 @@
+        )?)
+    };
+    // and report it to the user
+    println!("{}", location);
```
The `Display` implementations in `source.rs` produce human-readable output.  It'll be one of:

```
Copying image from /home/bgilbert/tmp/fedora-coreos-30.20191014.0-metal.x86_64.raw
Reading signature from /home/bgilbert/tmp/fedora-coreos-30.20191014.0-metal.x86_64.raw.sig
```
or:
```
Downloading image from http://localhost:8080/fedora-coreos-30.20191014.0-metal.x86_64.raw.xz
Downloading signature from http://localhost:8080/fedora-coreos-30.20191014.0-metal.x86_64.raw.xz.sig
```
or:
```
Downloading testing image and signature
```
or:
```
Downloading image and signature referenced from https://builds.coreos.fedoraproject.org/streams/testing.json
```

bgilbert commented on src/main.rs:
```diff
@@ -0,0 +1,464 @@
+        .setting(AppSettings::UnifiedHelpMessage)
+        // ways to specify the image source
+        .arg(
+            Arg::with_name("stream")
```
The current installer requires manually specifying a URL, on the command line or on the kernel command-line, and that's a really awkward UX.  I don't think users generally care, or should have to care, about the URL their OS image is coming from.  So I'd strongly prefer to keep the stream logic in coreos-installer, and also add something similar for RHCOS if there's equivalent metadata we can use.

By way of comparison, CL `coreos-install` understands channels, the `current` pointer, and `version.txt`.

lucab replied:
Oh yeah, I totally agree such logic should be in coreos-installer (I've just realized my previous comment was ambiguously worded, sorry).

My reasoning was focused on "how to expose this logic via CLI". Right now there are three switches that are mutually exclusive (plus additional options), and we likely have to fit RHCOS too in the future.

I was pondering alternative CLI approaches (still sharing the same logic under the hood), like:
 * dedicated verbs like `install-url`, `install-file`, `install-fcos-stream` and `install-rhcos-release`
 * composable verbs like `resolve-fcos-stream` and `resolve-rhcos-release`, producing URLs on stdout for other `install-url` and `fetch-url`

I hope it's a bit clearer now, sorry for the initial ambiguity.

bgilbert replied:
Ah, okay.  I agree that the current approach complicates the formal semantics, but it seems consistent with how other programs handle input/output options.  For that reason, the current approach feels more intuitive to me than having separate subcommands.  Also, to keep the UX simple, I'd rather not require piping.

There will also be a systemd unit that runs coreos-installer with parameters from kargs, and the kargs will necessarily have the same problem: there's not a great way to specify a mode.  We don't have to keep the two sets of options in sync (and in fact we can't, for compatibility reasons), but maybe it's still relevant here.

lucab replied:
Ack, I'll stick to your call here then.

bgilbert commented on src/main.rs:
```diff
@@ -0,0 +1,464 @@
+        let sniff = buf_reader.fill_buf().chain_err(|| "sniffing input")?;
+        if sniff.len() >= 2 && &sniff[0..2] == b"\x1f\x8b" {
+            Box::new(GzDecoder::new(buf_reader))
+        } else if sniff.len() >= 6 && &sniff[0..6] == b"\xfd7zXZ\x00" {
```
Yep, both of the magic numbers are endian-invariant.

bgilbert commented on src/source.rs:
```diff
@@ -0,0 +1,273 @@
+                .read_to_end(&mut signature)
+                .chain_err(|| "reading signature file")?;
+        } else {
+            eprintln!("Couldn't read signature file: {}", result.unwrap_err());
```
I was trying to avoid the extra `Option<_>` and had an explicit check for a zero-length vector, but you're right that I should do this properly.

lucab replied:
I acknowledge that the nested `Option<Vec<_>>` makes the code more annoying to write. On the upside, it makes clearer for any future reader what's going on.

bgilbert commented on src/blockdev.rs:
```diff
@@ -0,0 +1,168 @@
+        serde_json::from_slice(&*result.stdout).chain_err(|| "decoding lsblk JSON")?;
+    let mut found: Option<BlkDev> = None;
+    for dev in output.blockdevices {
+        if dev.label.is_none() || dev.label.as_ref().unwrap() != label {
```
Doesn't seem to.

lucab replied:
Uhm, sad.
I think this idiom will be covered by either [`contains()`](https://doc.rust-lang.org/std/option/enum.Option.html#method.contains) or [`as_deref()`](https://doc.rust-lang.org/std/option/enum.Option.html#method.as_deref) once those are stabilized.

bgilbert commented on src/main.rs:
```diff
@@ -0,0 +1,464 @@
+        .get_matches();
+
+    // build image location
+    let location: Box<dyn ImageLocation> = if matches.is_present("image-file") {
```
Yeah.  I'd prefer to use `ArgGroup`s and `conflicts_with`, but that doesn't seem to play well with Args that have defaults.

bgilbert commented:
Thanks for the review!  Updated, and Travis is enabled.  Also, as we discussed OOB, I've moved the functionality to an `install` subcommand for future-proofing.

lucab commented:
An additional wishlist item that came to my mind: it would be great if we could run the installer from a container, just with the bindmounts for the target (rationale: that would allow us to use any linux-host as a provisioner).
Did you maybe try that, or see any clear blocker? Does it sound useful to you (perhaps coupled with a quay autobuilder task) or am I getting off track?

jlebon commented on src/signing-keys.asc:
```diff
@@ -0,0 +1,91 @@
+-----BEGIN PGP PUBLIC KEY BLOCK-----
```
Hmm, why not just pick up the keys from `fedora-gpg-keys`/`redhat-release-coreos` (i.e. `/etc/pki/rpm-gpg`) ?

bgilbert replied:
coreos-installer should run on arbitrary distros, which may not have those files.

jlebon replied:
OK so the idea here is that we'll publish binaries instead of/in addition to containers? What I'm trying to avoid is adding yet another spot that needs updating during each major release rollover. Can we pull it in as part of the build process at least?

lucab replied:
It may make sense to overlay-merge (by filename?) two sets of keys:
 * these hardcoded ones (for fallback)
 * any on-disk ones (possibly fresher)

bgilbert replied:
> OK so the idea here is that we'll publish binaries instead of/in addition to containers?

It was, but if the container works and the UX for running it isn't too bad, I'd favor shipping that instead of a binary.

> any on-disk ones (possibly fresher)

Where would we get these?  If from `/etc/pki/rpm-gpg`, there are no symbolic references for the current key, so we'd have to either import everything we find there (boo) or have a heuristic for importing keys (boo).

If we do ship a bare binary, this would also mean that behavior would differ depending on the host OS, which doesn't seem great.

> What I'm trying to avoid is adding yet another spot that needs updating during each major release rollover. Can we pull it in as part of the build process at least?

These are the roots of trust for install images, and IMO it's safer to manage them manually.  We can always start that way and add dynamic roots or more automation later on.

In comparison, CL `coreos-install` intentionally hardcodes the GPG key and does not provide a command-line mechanism for overriding it.

jlebon replied:
So as an example right now we don't have any CI covering the `coreos-installer` path. So bumping from f30 to f31 without bumping the keys here would've possibly gone completely unnoticed until it shipped to users. (A gap we should fix of course!)

> Where would we get these? If from /etc/pki/rpm-gpg, there are no symbolic references for the current key, so we'd have to either import everything we find there (boo) or have a heuristic for importing keys (boo).

> These are the roots of trust for install images, and IMO it's safer to manage them manually. We can always start that way and add dynamic roots or more automation later on.

A big part of this discussion is essentially a direct consequence of https://github.com/coreos/fedora-coreos-tracker/issues/187#issuecomment-540103433. WDYT about:
- ship in a container with `fedora-gpg-keys`
- parse the FCOS version to extract the major release, similar to what is proposed for RoboSignatory itself in that ticket
- verify download using *only that release's specific key* from `/etc/pki/rpm-gpg`

?

This is similar to how Fedora verifies RPMs: https://src.fedoraproject.org/rpms/fedora-repos/blob/a12c809ab2dcb730cbeb4a03c50d20e3d25cc048/f/fedora.repo#_10.

For RHEL, the keys don't rotate nearly as much, so we can just bake it in.

Anyway, I'm cool with how things set up for now, and discussing this stuff in more details later on!

bgilbert replied:
> So as an example right now we don't have any CI covering the coreos-installer path. So bumping from f30 to f31 without bumping the keys here would've possibly gone completely unnoticed until it shipped to users. (A gap we should fix of course!)

If nothing else, that'll be addressed when we start running kola on Packet, where every machine will bootstrap with coreos-installer.

> ship in a container with fedora-gpg-keys

I'd still prefer not to rely on external keys.  It'd be great if `cargo install coreos-installer` worked properly on e.g. Debian.

>  parse the FCOS version to extract the major release, similar to what is proposed for RoboSignatory itself in that ticket
> verify download using only that release's specific key from /etc/pki/rpm-gpg

I like the idea of binding to a specific key.  Will file an issue.

Hmm, maybe stream metadata should list a signing key ID for each artifact?

bgilbert replied:
https://github.com/coreos/coreos-installer/issues/71

jlebon commented on src/main.rs:
```diff
@@ -0,0 +1,487 @@
+extern crate progress_streams;
+extern crate reqwest;
+extern crate tempdir;
+extern crate xz2;
```
Shouldn't need these with Rust 2018.

jlebon commented on src/main.rs:
```diff
@@ -0,0 +1,487 @@
+
+    // do the copy
+    config_dest.push("config.ign");
+    let mut config_in = OpenOptions::new()
```
Why not just `std::fs::copy` here?

bgilbert replied:
We get a bit better control this way:

- `config_out.create_new(true)` ensures that we'll fail if the file already exists for some reason.  (Which it shouldn't, of course.)
- `std::fs::copy` also copies permissions, which may not be desired.

jlebon commented on src/main.rs:
```diff
@@ -0,0 +1,487 @@
+            // Rewrite the config.  Assume that we will only install
+            // from metal images and that their bootloader configs will
+            // always set ignition.platform.id.  Fail if those
+            // assumptions change.  This is deliberately simplistic.
```
This is fine, though should probably no-op return early on if `platform` is `metal`.

bgilbert commented:
> it would be great if we could run the installer from a container, just with the bindmounts for the target (rationale: that would allow us to use any linux-host as a provisioner).
> Did you maybe try that, or see any clear blocker? Does it sound useful to you (perhaps coupled with a quay autobuilder task) or am I getting off track?

I haven't tried it, but I like the idea.  I was originally thinking we'd offer a binary for download, but it does have dynamic library dependencies so that may not work well on other distros:

```
	liblzma.so.5 => /lib64/liblzma.so.5 (0x00007fb47a786000)
	libssl.so.1.1 => /lib64/libssl.so.1.1 (0x00007fb47a6f0000)
	libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x00007fb47a414000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fb47a40e000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fb47a3ed000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fb47a3d3000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fb47a20b000)
	libz.so.1 => /lib64/libz.so.1 (0x00007fb47a1f1000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fb47ac29000)
```

A container image would sidestep the whole problem.  I probably won't pursue it immediately, but filed issue https://github.com/coreos/coreos-installer/issues/67.

bgilbert commented:
Updated!

lucab commented:
We could (in theory) cross-compile to a static binary with musl, but in practice if it works I'd prefer pursuing the container route.

crawford commented on src/verify.rs:
```diff
@@ -0,0 +1,142 @@
+
+use error_chain::bail;
+use std::fs::{metadata, set_permissions, OpenOptions};
+use std::io;
```
This can be combined below:

```rust
use std::io{self, Read, Write};
```

crawford commented on src/main.rs:
```diff
@@ -0,0 +1,480 @@
+    let uname = nix::sys::utsname::uname();
+    // Args are listed in --help in the order declared here.  Please keep
+    // the entire help text to 80 columns.
+    let app_matches = App::new("coreos-installer")
```
As an alternative to using Clap directly, StructOpt is pretty slick.

bgilbert replied:
Nice!  I played with it a bit and don't think it's worth a refactor right now, but I'll keep it in mind for the future.

bgilbert commented on src/main.rs:
```diff
@@ -0,0 +1,464 @@
+    insecure: bool,
+}
+
+quick_main!(run);
```
`quick_main` still does better pretty-printing than the default `main`.

lucab commented:
I don't have any further high-level concerns, and the code itself looks good. I could just stamp the PR but I don't know what are the overarching integration plans wrt. the old one.

@bgilbert @jlebon @arithx how do you want to proceed here?

jlebon commented:
Would be pretty cool to have the CI here actually do an install of FCOS. I think Travis CI can give you Ubuntu VMs, though we can also hook it up to CoreOS CI!

> @bgilbert @jlebon @arithx how do you want to proceed here?

Now that we've branched, I'm thinking let's just merge this, cut a 0.0.1 release, and iterate on top as we uncover issues?
